modules/pool/lib/pool-normalized-liquidty.ts

```diff
@@ -0,0 +1,315 @@
+                tokenPair.tokenA.address,
+                tokenPair.tokenB.address,
+                `${tokenPair.aToBAmountIn}`,
+                ZERO_ADDRESS,
+            ],
+            [ZERO_ADDRESS, false, ZERO_ADDRESS, false],
+        ],
+    );
+}
+
+// call querySwap from tokenA->tokenB with AtoB amount out
+function addBToAPriceCallsToMulticaller(
+    tokenPair: TokenPair,
+    balancerQueriesAddress: string,
+    multicaller: Multicaller3,
+) {
+    multicaller.call(
+        `${tokenPair.poolId}-${tokenPair.tokenA.address}-${tokenPair.tokenB.address}.bToAAmountOut`,
+        balancerQueriesAddress,
+        'querySwap',
+        [
+            [
+                tokenPair.poolId,
+                0,
+                tokenPair.tokenB.address,
+                tokenPair.tokenA.address,
+                `${tokenPair.aToBAmountOut}`,
+                ZERO_ADDRESS,
+            ],
+            [ZERO_ADDRESS, false, ZERO_ADDRESS, false],
+        ],
+    );
+}
+
+function getAmountOutAndEffectivePriceFromResult(tokenPair: TokenPair, onchainResults: { [id: string]: OnchainData }) {
+    const result = onchainResults[`${tokenPair.poolId}-${tokenPair.tokenA.address}-${tokenPair.tokenB.address}`];
+
+    if (result) {
+        tokenPair.aToBAmountOut = BigInt(result.aToBAmountOut.toString());
+        // MathSol expects all values with 18 decimals, need to scale them
+        tokenPair.effectivePrice = MathSol.divDownFixed(
+            parseUnits(tokenPair.effectivePriceAmountIn.toString(), 18 - tokenPair.tokenA.decimals),
+            parseUnits(result.effectivePriceAmountOut.toString(), 18 - tokenPair.tokenB.decimals),
+        );
+    }
+}
+
+function getBToAAmountFromResult(tokenPair: TokenPair, onchainResults: { [id: string]: OnchainData }) {
+    const result = onchainResults[`${tokenPair.poolId}-${tokenPair.tokenA.address}-${tokenPair.tokenB.address}`];
+
+    if (result) {
+        tokenPair.bToAAmountOut = BigInt(result.bToAAmountOut.toString());
+    }
+}
+function calculateSpotPrice(tokenPair: TokenPair) {
+    // MathSol expects all values with 18 decimals, need to scale them
+    const aToBAmountInScaled = parseUnits(tokenPair.aToBAmountIn.toString(), 18 - tokenPair.tokenA.decimals);
+    const aToBAmountOutScaled = parseUnits(tokenPair.aToBAmountOut.toString(), 18 - tokenPair.tokenB.decimals);
+    const bToAAmountOutScaled = parseUnits(tokenPair.bToAAmountOut.toString(), 18 - tokenPair.tokenA.decimals);
+    const priceAtoB = MathSol.divDownFixed(aToBAmountInScaled, aToBAmountOutScaled);
+    const priceBtoA = MathSol.divDownFixed(aToBAmountOutScaled, bToAAmountOutScaled);
+    tokenPair.spotPrice = MathSol.powDownFixed(MathSol.divDownFixed(priceAtoB, priceBtoA), WAD / 2n);
+}
+
+function calculateNormalizedLiquidity(tokenPair: TokenPair) {
+    // spotPrice and effective price are already scaled to 18 decimals by the MathSol output
+    const priceRatio = MathSol.divDownFixed(tokenPair.spotPrice, tokenPair.effectivePrice);
```

Comment:
Ideally you'd never get priceRatio above 1, but if that happens by any chance, you'll get negative price impact and normalized liquidity.
It would be nice to understand the implications of that in the path finding algorithm.
If that's an issue, you could set limits to results and enforce some behaviors.